source/_posts/2024-11-01-how-hooks-happened.md

```diff
@@ -0,0 +1,139 @@
+
+Eventually, Asymmetric visibility finally went down in a 14:12 vote in January of 2023.  RFCs need 2/3 majorities to pass, so 14:12, while a simple majority, isn't enough.
+
+## Hooking it up
+
+The failure of asymmetric visibility was a blow to both of us.  In most cases we didn't even know why the people who voted "no" did so, which made addressing their concerns guesswork.  Had we just wasted our time?  Would accessors even be acceptable?  The current RFC process is very bad at providing that kind of actionable feedback.
+
+While no one from the Foundation mentioned it, we were also both acutely aware that Ilija had been working on paid time for a proposal that ended up failing.
+
+Nonetheless, we had split the RFC in two deliberately so that even if only one was successful there was still a benefit.  We therefore turned our attention to the accessor part of the proposal (which included interface properties) and continued working.
+
+The model we developed was essentially a "direct port from Swift, with a few slight renames."
+
+The initial design included:
+
+* `get` - Totally takes over reading, and there's no physical property created.
+* `set` - Totally takes over writing, and there's no physical property created.
+* `beforeSet` - Intercepts the writing of a property but doesn't change the actual write itself.
+* `afterSet` - Called after the property is written, doesn't change the actual write itself.
+
+This is also when we changed the name from "accessors" to "hooks," which seemed a more accurate description given that model.  This later ended up not making sense, but Ilija had already taken the time to rename everything in the patch, so it was too much work to change back.  Sorry.
+
+The RFC text itself grew and grew.  PHP is a very mature (read: complex) language, and so there were lots of nooks and crannies that we had to account for.  Properties touch on almost everything: References, arrays, inheritance, final properties, interfaces, interaction with `readonly`, interaction with `__get`/`__set`, serialization, constructor property promotion, reflection...
+
+PHP is a big language, and we needed to think through and implement every possible edge case to avoid booby traps.
+
+## Some very good Advice
+
+We finally had a working design and implementation by April 2023 or so, but were quite nervous about it.  It was big, and we were both still smarting from the loss of asymmetric visibility.  Roman suggested that he run it past the PHP Foundation's Advisory Board for feedback first before going to the list, which turned out to be the best thing we could do.
+
+The Board came back with mostly positive feedback; they liked the concept, the design, the detail... but having 4 separate hooks felt very clumsy to them.  Especially with get/set wiping the backing property entirely and leaving the developer on their own.
+
+[Insert Foundation ad here]
```

Comment:
@pronskiy Go ahead and modify this however is appropriate for the corresponding Foundation commercial.